### PR TITLE
Removed node from multiarch buildbox, added i386 and amd64 targets

### PR DESCRIPTION
**Pull request by gzigzigzeo**
What's inside:
- [x] Changes required for GHA multistage/multiplatform build
- [x] Another fix for assets SHA calculation. Problem is that `yarn.lock` may change on `yarn install`. We calculate initial sha before `yarn install`, and save the resulting checksum with the changed `yarn.lock`. Therefore, the resulting checksum never matches the initial value.
- [x] Dockerfile-multiarch: Dockerfile-multiarch-base (gcc), Dockerfile-multiarch-clang (10 + 7 for boringcrypto), Dockerfile-multiarch
- [x] build.assets Makefile changes

### CODE REVIEW

File: build.assets/Dockerfile-multiarch [hunk 1]
```diff
@@ -44,6 +57,7 @@ ARG GOLANG_VERSION
 ARG RUST_VERSION
 ARG UID
 ARG GID
+ARG RUSTUP_VERSION=1.25.2
```
Comment (zmb3) [inline build.assets/Dockerfile-multiarch:60]: Do we need to pin the rustup version? Thus far we've pinned the Rust version, but used the latest version of rustup at time of install.

File: build.assets/Makefile [hunk 2]
```diff
@@ -446,18 +446,26 @@ define RELEASE_RECIPE
 			-v "/tmp:/tmp" \
 			-v "$(WORKSPACE):/go/teleport" \
 			-w /go/teleport \
-			-u 1000:1000 \
```
Comment (jakule) [inline build.assets/Makefile:449]: Why was this removed?

File: build.assets/build-webassets-if-changed.sh [hunk 3]
```diff
@@ -78,7 +78,8 @@ if [ "$BUILD" = "true" ]; then \
   # created any necessary directories here. The recalculation is necessary as yarn.lock may have been
   # updated by the build process.
   mkdir -p "$(dirname "$LAST_SHA_FILE")"
-  calculate_sha > "$LAST_SHA_FILE"
+  # Save SHA with yarn.lock before yarn install
+  echo $CURRENT_SHA > "$LAST_SHA_FILE"
```
Comment (jakule) [inline build.assets/build-webassets-if-changed.sh:82]: 🙌

File: build.assets/Dockerfile-multiarch [hunk 4]
```diff
@@ -25,12 +27,23 @@ RUN echo "armhfp" > /etc/yum/vars/basearch && \
...
 
+## 386 ########################################################################
+FROM base AS platform-setup-386
+ARG GOLANG_ARCH=386
+ARG RUST_ARCH=i686-unknown-linux-gnu
+
+## AMD 64 #####################################################################
+FROM base AS platform-setup-amd64
+ARG GOLANG_ARCH=amd64
+ARG RUST_ARCH=x86_64-unknown-linux-gnu
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch:45]: Why do we need 386/AMD64 here? We're using different docker images for that.

File: build.assets/Dockerfile-multiarch (hunk 1 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch:60]: Sure. Sorry, that's a leftover from an experiment, removed.

File: build.assets/Dockerfile-multiarch (hunk 4 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch:45]: That was the original idea for this specific Dockerfile: to use a single build image to build for all platforms. I guess, I am a little late replying you here, and we've already discussed it in DM. Just to Drone builds minimal images, so, let me leave and use it here for Drone and we'll sync about again later.

File: build.assets/Makefile (hunk 2 above)
Comment (gzigzigzeo) [reply build.assets/Makefile:449]: I had the problems with local fs permissions. 1000:1000 did not match the local user/group. You know, let me discard all changes to Makefile here. I guess, those docker flags should be on the GHA side anyway.

File: .github/workflows/build-ci-gcc-centos7.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    paths:
+      - .github/services/Dockerfile.*
+    branches:
+      - master
+  pull_request:
+
+env:
+  # REGISTRY: ghcr.io
+  IMAGE_NAME: gzigzigzeo/gcc-centos7
+  GCC_VERSION: 4.9.4
```
Comment (jakule) [inline .github/workflows/build-ci-gcc-centos7.yaml:14]: nit: Why are we using GCC 4.9? If we're building from the source, we can use 12.2.

File: build.assets/Dockerfile-multiarch [hunk 6]
```diff
@@ -25,12 +27,88 @@ RUN echo "armhfp" > /etc/yum/vars/basearch && \
...
+
+## LIBPCSCLITE ################################################################
+# 
+FROM ghcr.io/gravitational/buildbox-multiarch-devtoolset:teleport13-gcc-$TARGETARCH AS libpcsclite
+
+# TODO: -> MAKEFILE (?), upper level (?)
+ARG LIBPCSCLITE_VERSION=1.9.9-teleport 
+
+ENV CFLAGS="-std=gnu99 -latomic"
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch:54]: AFAIR `libatomic` is a shared library we cannot expect on the target system.

File: build.assets/Dockerfile-multiarch-devtoolset-gcc [hunk 7]
```diff
@@ -0,0 +1,98 @@
+FROM centos:7 AS base
+
+# Automatically supplied by the Docker buildkit
+ARG TARGETARCH
+
+# Minimal build tool versions
+ARG GCC_VERSION=7.4.0
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-devtoolset-gcc:7]: The same as above, we can probably switch to something never, but if this is going to take too much time, I don't mind the current version.

File: build.assets/Dockerfile-multiarch-devtoolset-gcc [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+FROM platform-setup-$TARGETARCH
+
+# Prerequestiques
+RUN yum update -y $YUM_UPDATE_FLAGS && \
+    yum install -y \
+        automake \
+        libtool \
+        flex \        
+        autoconf-archive \
+        git \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-devtoolset-gcc:52]: The git from the centos repo doesn't work in GHA (v2.x is needed to support `git submodules`)

File: build.assets/Makefile [hunk 9]
```diff
@@ -554,3 +554,48 @@ build-centos7-assets:
...
 		--tag $(BUILDBOX_CENTOS7_ASSETS)-$(RUNTIME_ARCH) \
 		-f Dockerfile-centos7-assets .
+
+#################################################################################
+# Multiarch build steps
+#################################################################################
+
+GCC_VERSION ?= 10.4.0
+CLANG_10_VERSION ?= 10.0.1
+CLANG_7_VERSION ?= 7.0.1
```
Comment (jakule) [inline build.assets/Makefile:564]: nit: I don't think we need this as a Makefiles variable. I don't think we will update it any time soon.

File: build.assets/Dockerfile-multiarch-devtoolset-gcc (hunk 8 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch-devtoolset-gcc:58]: We do not need the newer git inside the container anymore. Cloning is done by the GHA themselves on host OS.

File: build.assets/Dockerfile-multiarch-base [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        gcc-c++ \
+        glibc-static \
+        glibc-devel \
+        libstdc++-static \
+        libstdc++-devel \
+        bzip2 \
+        file \
+        binutils \
+        zlib-devel \
+        elfutils-libelf-devel-static \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-base:93]: I don't think that `elfutils-libelf-devel-static` is needed here

File: build.assets/Dockerfile-multiarch-base [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+RUN yum update -y $YUM_UPDATE_FLAGS && \
+    yum install -y     \
+        git \
+        gcc \
+        gcc-c++ \
+        make \
+        pam-devel \
+        perl-IPC-Cmd \
+        tree \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-base:72]: `tree` is not needed here

File: build.assets/Dockerfile-multiarch-base [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+ENV LANGUAGE=en_US.UTF-8 \
+    LANG=en_US.UTF-8 \
+    LC_ALL=en_US.UTF-8 \
+    LC_CTYPE=en_US.UTF-8
+
+RUN yum update -y $YUM_UPDATE_FLAGS && \
+    yum install -y     \
+        git \
+        gcc \
+        gcc-c++ \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-base:68]: Theses dependencies are listed twice. Can you sort them by name and remove duplicates?

File: build.assets/Dockerfile-multiarch-base [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+FROM platform-setup-$TARGETARCH AS gcc
+
+ENV LANGUAGE=en_US.UTF-8 \
+    LANG=en_US.UTF-8 \
+    LC_ALL=en_US.UTF-8 \
+    LC_CTYPE=en_US.UTF-8
+
+RUN yum update -y $YUM_UPDATE_FLAGS && \
+    yum install -y     \
+        git \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-base:66]: I don't see `git` being used anywhere

File: build.assets/Dockerfile-multiarch-base [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    echo "armv7hl-redhat-linux-gnu" > /etc/rpm/platform
+
+## ARM 64 #####################################################################
+FROM base AS platform-setup-arm64
+ENV GCC_BUILD_FLAGS="--build aarch64-unknown-linux-gnu"
+ENV GOLANG_ARCH=arm64
+ENV RUST_ARCH=aarch64-unknown-linux-gnu
+
+# Installing the kerenel packages causes the update to hang on aarch64, so we
+# skip upgrading them.
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-base:40]: Are you sure this is the case? AFAIR that was only a problem when running with QEMU. Is this is still the case here?

File: build.assets/Dockerfile-multiarch [hunk 15]
```diff
@@ -50,25 +26,213 @@ RUN groupadd ci --gid=$GID -o && \
...
+    make && \
+    make install
+
+# Install libfido2.
+# Depends on libcbor, openssl, zlib-devel and libudev.
+# Linked so `make build/tsh` finds the library where it expects it.
+RUN git clone --depth=1 https://github.com/Yubico/libfido2.git -b 1.12.0 && \
+    cd libfido2 && \
+    [ "$(git rev-parse HEAD)" = '659a02679f99fd34a44e06e35dce90794f6ecc86' ] && \
+      LDFLAGS="-Wl,--copy-dt-needed-entries" cmake3 \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch:85]: Why `--copy-dt-needed-entries` is needed here? I don't see that is the case on master.

CC @codingllama

File: build.assets/Dockerfile-multiarch [hunk 16]
```diff
@@ -50,25 +26,213 @@ RUN groupadd ci --gid=$GID -o && \
...
-        zip \
-        zlib-static && \
-    yum clean all && \
-    localedef -c -i en_US -f UTF-8 en_US.UTF-8
+## LIBPCSCLITE ################################################################
+# 
+FROM gcc AS libpcsclite
+ARG LIBPCSCLITE_VERSION
+
+ENV CFLAGS="-std=gnu99"
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch:34]: Why is this flag needed? I'd expect this to be set by  `configure`.

File: build.assets/Dockerfile-multiarch [hunk 17]
```diff
@@ -50,25 +26,213 @@ RUN groupadd ci --gid=$GID -o && \
...
+    cd build && \
+    cd /opt/boringssl/build && cmake3 -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DFIPS=1 -DCMAKE_BUILD_TYPE=Release -GNinja .. && ninja
+
+## Intermediate targets ########################################################
+#
+FROM gcc AS deps-arm
+
+ENV GOARCH=arm
+ENV GOARM=7
+# ENV GOEXPERIMENT=nounified (?) https://github.com/golang/go/issues/58425
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch:213]: Leftover comment?

File: build.assets/Dockerfile-multiarch [hunk 18]
```diff
@@ -50,25 +26,213 @@ RUN groupadd ci --gid=$GID -o && \
...
+
+RUN mkdir -p $RUSTUP_HOME && chmod a+w $RUSTUP_HOME && \
+    mkdir -p $CARGO_HOME/registry && chmod -R a+w $CARGO_HOME
+
+USER ci
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain ${RUST_VERSION} --default-host ${RUST_ARCH} && \
+    rustup --version && \
+    cargo --version && \
+    rustc --version && \
+    rustup component add rustfmt clippy && \
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch:130]: We don't need `rustfmt` and `clippy` in our release images.
```suggestion
```

File: build.assets/Dockerfile-multiarch-base [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        bzip2 \
+        file \
+        binutils \
+        zlib-devel \
+        elfutils-libelf-devel-static \
+        texinfo && \
+    yum clean all && \
+    localedef -c -i en_US -f UTF-8 en_US.UTF-8
+
+# Compile & install GCC
```
Comment (jakule) [inline build.assets/Dockerfile-multiarch-base:98]: From what I see, we're compiling GCC 10 against the dependencies used to build GCC 4.8. Is this an officially supported combination, or does it just happen to work? I'm talking about MPFR, MPC, etc.

File: build.assets/Makefile [hunk 20]
```diff
@@ -554,3 +562,74 @@ build-centos7-assets:
...
```
Comment (jakule) [inline build.assets/Makefile:572]: Can we rename these variables to suggest usage rather than version?
```suggestion
CLANG_BPF_VERSION ?= 10.0.1
CLANG_FIPS_VERSION ?= 7.0.1
```

File: build.assets/Dockerfile-multiarch (hunk 15 above)
Comment (codingllama) [reply build.assets/Dockerfile-multiarch:85]: Beats me too.

File: build.assets/Dockerfile-multiarch-base (hunk 13 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch-base:66]: It is used to clone clang, libudev, etc.

File: build.assets/Dockerfile-multiarch-base (hunk 10 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch-base]: 2023-03-02T10:55:04.2019152Z #31 1.247 Package libelf was not found in the pkg-config search path.
2023-03-02T10:55:04.2020851Z #31 1.247 Perhaps you should add the directory containing `libelf.pc'
2023-03-02T10:55:04.2021739Z #31 1.247 to the PKG_CONFIG_PATH environment variable

File: build.assets/Dockerfile-multiarch-base (hunk 14 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch-base:39]: No, I left it as is from the original Dockerfile-multiarch. Everything works without it. Removed.

File: build.assets/Dockerfile-multiarch-base (hunk 19 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch-base:95]: Yes. It depends. Some deps require newer GCC to build - but as far as I understood, we meet minimal versions for our case (4.8.3+).

File: build.assets/Dockerfile-multiarch (hunk 16 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch:34]: Yes, but on ARM it sets the wrong value for some reason. Had to override it here.

File: build.assets/Dockerfile-multiarch (hunk 15 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch]: #20 22.74 /usr/bin/ld: /usr/local/lib/gcc/x86_64-unknown-linux-gnu/10.4.0/../../../../lib64/libudev.a(udev_enumerate.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
#20 22.74 //usr/lib64/libpthread.so.0: error adding symbols: DSO missing from command line
(ARM64)

> With --copy-dt-needed-entries dynamic libraries mentioned on the command line will be recursively searched, following their DT_NEEDED tags to other libraries, in order to resolve symbols required by the output binary. With the default setting however the searching of dynamic libraries that follow it will stop with the dynamic library itself. No DT_NEEDED links will be traversed to resolve symbols.

File: build.assets/Dockerfile-multiarch (hunk 16 above)
Comment (jakule) [reply build.assets/Dockerfile-multiarch:35]: Can you please add a comment that this is needed on ARM? I see someone testing it on x86 and removing the comment as "everything works".

File: build.assets/Dockerfile-multiarch-base (hunk 10 above)
Comment (jakule) [reply build.assets/Dockerfile-multiarch-base]: Where is this log come from? AFAIK `libelf` is only needed by `libbpf`.

File: build.assets/Dockerfile-multiarch (hunk 15 above)
Comment (jakule) [reply build.assets/Dockerfile-multiarch:86]: Sounds like a wrong linking order. Can you try to append `-lpthread` instead of "recursive search". Sounds like only this one library is missing and this recursive search can do some damage in rare cases.

File: build.assets/Dockerfile-multiarch (hunk 16 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch:35]: Yep, already done.

File: build.assets/Dockerfile-multiarch-base (hunk 10 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch-base]: From libbpf build log. I can install it separately at the libbpf build stage, but I guess it is better to keep all deps in one place.

File: build.assets/Dockerfile-multiarch (hunk 15 above)
Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch:86]: Okay, will try.

Comment (gzigzigzeo) [reply build.assets/Dockerfile-multiarch]: Yes, you were correct. It works with -lpthread, my fault here.

**Review by jakule — APPROVED** [review]

**Comment by jakule** [comment]
@r0mant @zmb3 @tcsc Friendly ping.

File: Makefile [hunk 21]
```diff
@@ -208,7 +208,12 @@ ifeq ("$(ARCH)","arm64")
 	endif
 else ifeq ("$(ARCH)","arm")
 # ARM builds need to specify the correct C compiler
```
Comment (zmb3) [inline Makefile:210]: This comment no longer makes sense. Move it down to the `CC=` line and update so it says `Cross compilation must specify the correct C compiler`.

**Review by zmb3 — APPROVED** [review]

**Comment by jakule** [comment]
Merged as https://github.com/gravitational/teleport/pull/22979
